nimbus-consensus/src/lib.rs

```diff
@@ -153,6 +156,125 @@ where
 	}
 }
 
+/// Grabs any available nimbus key from the keystore.
+/// This may be useful in situations where you expect exactly one key
+/// and intend to perform an operation with it regardless of whether it is
+/// expected to be eligible. Concretely, this is used in the consensus worker
+/// to implement the `skip_prediction` feature.
+pub(crate) fn first_available_key(keystore: &dyn SyncCryptoStore) -> Option<CryptoTypePublicPair> {
+	// Get all the available keys
+	let available_keys =
+		SyncCryptoStore::keys(keystore, NIMBUS_KEY_ID)
+		.expect("keystore should return the keys it has");
+
+	// Print a more helpful message than "not eligible" when there are no keys at all.
+	if available_keys.is_empty() {
+		warn!(target: LOG_TARGET, "🔏 No Nimbus keys available. We will not be able to author.");
+		return None;
+	}
+
+	Some(available_keys[0].clone())
+}
+
+/// Grab the first eligible nimbus key from the keystore
+/// If multiple keys are eligible this function still only returns one
+/// and makes no guarantees which one as that depends on the keystore's iterator behavior.
+/// This is the standard way of determining which key to author with.
+pub(crate) fn first_eligible_key<B: BlockT, C>(client: Arc<C>, keystore: &dyn SyncCryptoStore, parent: &B::Header, slot_number: u32) -> Option<CryptoTypePublicPair>
+where
+	C: ProvideRuntimeApi<B>,
+	C::Api: AuthorFilterAPI<B, NimbusId>,
+{
+	// Get all the available keys
+	let available_keys =
+		SyncCryptoStore::keys(keystore, NIMBUS_KEY_ID)
+		.expect("keystore should return the keys it has");
+
+	// Print a more helpful message than "not eligible" when there are no keys at all.
+	if available_keys.is_empty() {
+		warn!(target: LOG_TARGET, "🔏 No Nimbus keys available. We will not be able to author.");
+		return None;
+	}
+
+	// Get `AuthorFilterAPI` version.
+	let api_version = client.runtime_api()
+		.api_version::<dyn AuthorFilterAPI<B, NimbusId>>(&BlockId::Hash(parent.hash()))
```

Comment:
NimbusAPI should be used instead if it exists, like in the other PR.

Reply:
Good catch. I think I resolved the merge incorrectly.